src/components/shared/ToggleVisibilityButton.tsx

```diff
@@ -0,0 +1,63 @@
+import { ButtonBase, CircularProgress } from "@mui/material";
+import { Visibility, VisibilityOff } from "@mui/icons-material";
+import { useEffect, useState } from "react";
+
+export interface HideOracleButtonProps {
+  onToggleVisibility: () => void;
+  loading: boolean;
+  hidden?: boolean;
+}
+
+export function ToggleVisibilityButton(
+  props: HideOracleButtonProps
+) {
+  const { onToggleVisibility, loading, hidden } = props;
+
+  const [buttonLoading, setButtonLoading] = useState(false);
+
+  useEffect(() => {
+    if(!loading && buttonLoading) {
+      setButtonLoading(false);
+    }
+    if(loading && !buttonLoading) {
+      setButtonLoading(true);
+    }
+  }, [buttonLoading, loading]);
+
+  return (
+    <ButtonBase
+      onClick={() => {
+        setButtonLoading(true);
+        onToggleVisibility();
+      }}
+      disabled={loading}
+      disableRipple
+      aria-label={"Toggle visibility"}
```

Comment:
Everything looks really good - only change I would request would be adding the current state to this aria-label. Right now I don't think users using screen readers would be able to tell which oracles are visible and which ones are hidden. Adding "Hide Item" and "Show Item" here should fix that.